### PR TITLE
Include what you use

### PR DESCRIPTION
**Pull request by omartijn**
This PR adds the `iwyu` target to CMakeLists.txt, which, when run, executed the iwyu_tool, checking for missing or extraneous includes.

By making a contribution to this project, I certify that:

        (a) The contribution was created in whole or in part by me and I
            have the right to submit it under the GPL-3.0 license; or

        (b) The contribution is based upon previous work that, to the best
            of my knowledge, is covered under an appropriate open source
            license and I have the right under that license to submit that
            work with modifications, whether created in whole or in part
            by me, under GPL-3.0 license; or

        (c) The contribution was provided directly to me by some other
            person who certified (a), (b) or (c) and I have not modified
            it.

        (d) I understand and agree that this project and the contribution
            are public and that a record of the contribution (including all
            personal information I submit with it, including my sign-off) is
            maintained indefinitely and may be redistributed consistent with
            this project or the license(s) involved.

### CODE REVIEW

**Comment by tomsmeding** [comment]
> I'm guessing the comments on the `#include`'s were placed there by iwyu? Unless we add those comments for _every_ `#include` in the source, I think we should remove them. Partly because a lot of them just add noise (e.g. `#include <vector> // for vector`), and partly because they'll be a pain to maintain unless we regularly run the tool on the source.

I thing I agree with James here; such comments are sometimes useful to explain why a particular obscure header file is necessary, but having them for _all_ includes seems unnecessary, and adds an additional maintenance burden too.

File: CMakeLists.txt [hunk 1]
```diff
@@ -161,3 +163,13 @@ install(
     DESTINATION
         lib/cmake/pgp-packet
 )
+
+find_package(PythonInterp)
+find_program(iwyu_tool_path NAMES iwyu_tool.py iwyu_tool)
+if (iwyu_tool_path AND PYTHONINTERP_FOUND)
+    add_custom_target(iwyu
+        COMMAND "${PYTHON_EXECUTABLE}" "${iwyu_tool_path}" -p "${CMAKE_BINARY_DIR}" -- --no_fwd_decls
```
[the remark this answers is not in the document]
Comment (omartijn) [reply CMakeLists.txt]: That's an option. Ideally we'd want the target to not appear if no suitable iwyu_tool is found. I could also remove the --no_fwd_decl flag, which should make it compatible with older versions. I'll see if that gives sensible results.

File: source/multiprecision_integer.cpp [hunk 2]
```diff
@@ -104,8 +108,8 @@ namespace pgp {
     {
         // erase any leading zero bytes
         data.erase(
-            std::begin(data),
-            std::find_if(std::begin(data), std::end(data), [](uint8_t b) { return b != 0; })
+            begin(data),
+            std::find_if(begin(data), end(data), [](uint8_t b) { return b != 0; })
```
[the remark this answers is not in the document]
Comment (omartijn) [reply source/multiprecision_integer.cpp]: We could. Is there a particular reason you'd want that?

**Comment by omartijn** [comment]
@JamesWrigley @tomsmeding Those include comments indeed originate from iwyu. They are not always directly useful. The `<vector>` include - as mentioned - is a good example of this.

It's true that you can say it generates noise when done for every include. I personally think that it is preferably over only doing it for _some_, since having it everywhere creates some kind of uniformity.

If we _do_ want it only for the more obscure includes, how do we qualify obscure? Is including `<utility>` to get `std::move` obscure? Technically you have to do that, but nobody does, since practically _every_ header in the STL does include it.

**Comment by tomsmeding** [comment]
@omartijn For what names would you say it is clearly useful?

**Comment by JamesWrigley** [comment]
I don't think having the comments adds any value though? Like, if I saw `#include <utility>` in there and suspected that it wasn't necessary, I could now just run `make iwyu` and know for sure.

I agree that it's preferable to doing it for all of them instead of for only some, as you say the lines do get a bit blurry when deciding what is/isn't obscure. But for maintainability this should be run before every commit, or at least semi-regularly (e.g., a pre-commit hook).

**Comment by omartijn** [comment]
Well, I personally found the realisation that I need to include `<utility>` surprising. It's also a good point that ideally we'd have a pre-commit hook which checks the includes. Unfortunately, the iwyu tool is far from infallible and comes with some incorrect suggestions, so we couldn't possible make a commit fail on this.

It's indeed true that it's a good idea to run iwyu every time you make some change to see whether what you did was correct. In that case the comments are indeed not necessary, so I'll remove them.

**Pull request merged**